File: crates/hologram-ffi/interfaces/python/examples/performance_benchmarks.py
#!/usr/bin/env python3
"""
Performance Benchmarks

This example demonstrates performance testing and benchmarking including:
- Memory allocation performance
- Buffer operation performance
- Tensor operation performance
- Atlas state management performance
- Performance comparison and analysis
"""

import hologram_ffi as hg
import json
import time
import statistics

def benchmark_executor_creation():
    """Benchmark executor creation performance."""
    print("\n🏗️ Benchmarking Executor Creation:")
    
    times = []
    for i in range(10):
        start_time = time.time()
        executor_handle = hg.new_executor()
        end_time = time.time()
        
        creation_time = (end_time - start_time) * 1000  # Convert to milliseconds
        times.append(creation_time)
        
        # Cleanup
        hg.executor_cleanup(executor_handle)
    
    avg_time = statistics.mean(times)
    min_time = min(times)
    max_time = max(times)
    std_time = statistics.stdev(times) if len(times) > 1 else 0
    
    print(f"  Average: {avg_time:.2f} ms")
    print(f"  Min: {min_time:.2f} ms")
    print(f"  Max: {max_time:.2f} ms")
    print(f"  Std Dev: {std_time:.2f} ms")
    
    return avg_time

def benchmark_buffer_operations():
    """Benchmark buffer operation performance."""
    print("\n📦 Benchmarking Buffer Operations:")
    
    executor_handle = hg.new_executor()
    
    try:
        # Benchmark buffer allocation
        allocation_times = []
        buffer_handles = []
        
        for i in range(100):
            start_time = time.time()
            buffer_handle = hg.executor_allocate_buffer(executor_handle, 1000)
            end_time = time.time()
            
            allocation_time = (end_time - start_time) * 1000
            allocation_times.append(allocation_time)
            buffer_handles.append(buffer_handle)
        
        avg_allocation = statistics.mean(allocation_times)
        print(f"  Buffer Allocation (1000 elements):")
        print(f"    Average: {avg_allocation:.2f} ms")
        
        # Benchmark buffer operations
        if buffer_handles:
            # Fill operation
            fill_times = []
            for buffer_handle in buffer_handles[:10]:  # Test first 10
                start_time = time.time()
                hg.buffer_fill(buffer_handle, 42.0, 1000)
                end_time = time.time()
                
                fill_time = (end_time - start_time) * 1000
                fill_times.append(fill_time)
            
            avg_fill = statistics.mean(fill_times)
            print(f"  Buffer Fill (1000 elements):")
            print(f"    Average: {avg_fill:.2f} ms")
            
            # Copy operation
            copy_times = []
            for i in range(10):
                if i + 1 < len(buffer_handles):
                    start_time = time.time()
                    hg.buffer_copy(buffer_handles[i], buffer_handles[i + 1], 1000)
                    end_time = time.time()
                    
                    copy_time = (end_time - start_time) * 1000
                    copy_times.append(copy_time)
            
            avg_copy = statistics.mean(copy_times)
            print(f"  Buffer Copy (1000 elements):")
            print(f"    Average: {avg_copy:.2f} ms")
        
        # Cleanup buffers
        for buffer_handle in buffer_handles:
            hg.buffer_cleanup(buffer_handle)
        
        return avg_allocation
        
    finally:
        hg.executor_cleanup(executor_handle)

def benchmark_tensor_operations():
    """Benchmark tensor operation performance."""
    print("\n📊 Benchmarking Tensor Operations:")
    
    executor_handle = hg.new_executor()
    
    try:
        # Benchmark tensor operations
        operations = [
            ("Reshape", lambda: _benchmark_tensor_reshape(executor_handle)),
            ("Transpose", lambda: _benchmark_tensor_transpose(executor_handle)),
            ("View 1D", lambda: _benchmark_tensor_view_1d(executor_handle)),
        ]
        
        for op_name, op_func in operations:
            try:
                avg_time = op_func()
                print(f"  {op_name}: {avg_time:.2f} ms")
            except Exception as e:
                print(f"  {op_name}: Failed - {e}")
        
    finally:
        hg.executor_cleanup(executor_handle)

def _benchmark_tensor_reshape(executor_handle):
    """Benchmark tensor reshape operation."""
    times = []
    for i in range(10):
        # Create fresh buffer and tensor for each iteration
        buffer_handle = hg.executor_allocate_buffer(executor_handle, 10000)  # 10K elements
        tensor_handle = hg.tensor_from_buffer(buffer_handle, json.dumps([100, 100]))  # 100x100 matrix
        
        start_time = time.time()
        result_handle = hg.tensor_reshape(tensor_handle, json.dumps([1000, 10]))
        end_time = time.time()
        
        op_time = (end_time - start_time) * 1000
        times.append(op_time)
        
        # Cleanup
        hg.tensor_cleanup(result_handle)
    
    return statistics.mean(times)

def _benchmark_tensor_transpose(executor_handle):
    """Benchmark tensor transpose operation."""
    times = []
    for i in range(10):
        # Create fresh buffer and tensor for each iteration
        buffer_handle = hg.executor_allocate_buffer(executor_handle, 10000)  # 10K elements
        tensor_handle = hg.tensor_from_buffer(buffer_handle, json.dumps([100, 100]))  # 100x100 matrix
        
        start_time = time.time()
        result_handle = hg.tensor_transpose(tensor_handle)
        end_time = time.time()
        
        op_time = (end_time - start_time) * 1000
        times.append(op_time)
        
        # Cleanup
        hg.tensor_cleanup(result_handle)
    
    return statistics.mean(times)

def _benchmark_tensor_view_1d(executor_handle):
    """Benchmark tensor view_1d operation."""
    times = []
    for i in range(10):
        # Create fresh buffer and tensor for each iteration
        buffer_handle = hg.executor_allocate_buffer(executor_handle, 10000)  # 10K elements
        tensor_handle = hg.tensor_from_buffer(buffer_handle, json.dumps([100, 100]))  # 100x100 matrix
        
        start_time = time.time()
        result_handle = hg.tensor_view_1d(tensor_handle)
        end_time = time.time()
        
        op_time = (end_time - start_time) * 1000
        times.append(op_time)
        
        # Cleanup
        hg.tensor_cleanup(result_handle)
    
    return statistics.mean(times)

def benchmark_atlas_operations():
    """Benchmark Atlas state management performance."""
    print("\n🌐 Benchmarking Atlas Operations:")
    
    # Benchmark phase operations
    phase_times = []
    for i in range(1000):
        start_time = time.time()
        phase = hg.atlas_phase()
        end_time = time.time()
        
        phase_time = (end_time - start_time) * 1000
        phase_times.append(phase_time)
    
    avg_phase = statistics.mean(phase_times)
    print(f"  Atlas Phase (1000 calls):")
    print(f"    Average: {avg_phase:.4f} ms")
    
    # Benchmark resonance operations
    resonance_times = []
    for i in range(100):
        start_time = time.time()
        resonance = hg.atlas_resonance_at(i % 96)  # Valid class range
        end_time = time.time()
        
        resonance_time = (end_time - start_time) * 1000
        resonance_times.append(resonance_time)
    
    avg_resonance = statistics.mean(resonance_times)
    print(f"  Atlas Resonance (100 calls):")
    print(f"    Average: {avg_resonance:.4f} ms")
    
    # Benchmark snapshot operation
    snapshot_times = []
    for i in range(10):
        start_time = time.time()
        snapshot = hg.atlas_resonance_snapshot()
        end_time = time.time()
        
        snapshot_time = (end_time - start_time) * 1000
        snapshot_times.append(snapshot_time)
    
    avg_snapshot = statistics.mean(snapshot_times)
    print(f"  Atlas Snapshot (10 calls):")
    print(f"    Average: {avg_snapshot:.2f} ms")
    
    return avg_phase, avg_resonance, avg_snapshot

def benchmark_memory_usage():
    """Benchmark memory usage patterns."""
    print("\n💾 Benchmarking Memory Usage:")
    
    executor_handle = hg.new_executor()
    
    try:
        # Test memory allocation patterns
        buffer_sizes = [100, 1000, 10000, 100000]
        
        for size in buffer_sizes:
            times = []
            handles = []
            
            # Allocate multiple buffers of same size
            for i in range(5):
                start_time = time.time()
                buffer_handle = hg.executor_allocate_buffer(executor_handle, size)
                end_time = time.time()
                
                allocation_time = (end_time - start_time) * 1000
                times.append(allocation_time)
                handles.append(buffer_handle)
            
            avg_time = statistics.mean(times)
            print(f"  Buffer Size {size}: {avg_time:.2f} ms average")
            
            # Cleanup
            for handle in handles:
                hg.buffer_cleanup(handle)
        
    finally:
        hg.executor_cleanup(executor_handle)

def main():
    print("=" * 60)
    print("Hologram FFI - Performance Benchmarks")
    print("=" * 60)
    
    try:
        # Run all benchmarks
        executor_time = benchmark_executor_creation()
        buffer_time = benchmark_buffer_operations()
        benchmark_tensor_operations()
        atlas_times = benchmark_atlas_operations()
        benchmark_memory_usage()
        
        # Summary
        print("\n📊 Performance Summary:")
        print(f"  Executor Creation: {executor_time:.2f} ms")
        print(f"  Buffer Allocation: {buffer_time:.2f} ms")
        print(f"  Atlas Phase: {atlas_times[0]:.4f} ms")
        print(f"  Atlas Resonance: {atlas_times[1]:.4f} ms")
        print(f"  Atlas Snapshot: {atlas_times[2]:.2f} ms")
        
        print("\n✅ Performance benchmarks completed!")
        
    except Exception as e:
        print(f"❌ Benchmark failed: {e}")

if __name__ == "__main__":
    main()

File: crates/hologram-ffi/interfaces/typescript/examples/performance_benchmarks.ts
/**
 * Performance Benchmarks Example
 * 
 * This example demonstrates performance testing and benchmarking of various
 * hologram-ffi operations to measure overhead and performance characteristics.
 */

import * as hg from '../src/index';

console.log('🚀 Hologram FFI - Performance Benchmarks Example');
console.log('===============================================');

// Benchmark configuration
const BENCHMARK_CONFIG = {
  iterations: 1000,
  warmupIterations: 100,
  bufferSizes: [100, 1000, 10000, 100000],
  tensorShapes: [
    [10, 10],
    [100, 100],
    [1000, 100],
    [100, 1000]
  ]
};

// Utility functions
function measureTime<T>(fn: () => T): { result: T; time: number } {
  const start = process.hrtime.bigint();
  const result = fn();
  const end = process.hrtime.bigint();
  const time = Number(end - start) / 1000000; // Convert to milliseconds
  return { result, time };
}

function benchmark(name: string, fn: () => void, iterations: number = BENCHMARK_CONFIG.iterations): void {
  console.log(`\n📊 Benchmarking: ${name}`);
  
  // Warmup
  for (let i = 0; i < BENCHMARK_CONFIG.warmupIterations; i++) {
    fn();
  }
  
  // Actual benchmark
  const times: number[] = [];
  for (let i = 0; i < iterations; i++) {
    const { time } = measureTime(fn);
    times.push(time);
  }
  
  // Calculate statistics
  times.sort((a, b) => a - b);
  const min = times[0] || 0;
  const max = times[times.length - 1] || 0;
  const median = times[Math.floor(times.length / 2)] || 0;
  const mean = times.reduce((sum, time) => sum + time, 0) / times.length;
  const p95 = times[Math.floor(times.length * 0.95)] || 0;
  const p99 = times[Math.floor(times.length * 0.99)] || 0;
  
  console.log(`  Iterations: ${iterations}`);
  console.log(`  Min: ${min.toFixed(3)}ms`);
  console.log(`  Max: ${max.toFixed(3)}ms`);
  console.log(`  Mean: ${mean.toFixed(3)}ms`);
  console.log(`  Median: ${median.toFixed(3)}ms`);
  console.log(`  P95: ${p95.toFixed(3)}ms`);
  console.log(`  P99: ${p99.toFixed(3)}ms`);
  console.log(`  Ops/sec: ${(1000 / mean).toFixed(0)}`);
}

// Core function benchmarks
console.log('\n🔧 Core Function Benchmarks:');

benchmark('getVersion()', () => {
  hg.getVersion();
});

benchmark('getExecutorPhase()', () => {
  hg.getExecutorPhase();
});

benchmark('advanceExecutorPhase(1)', () => {
  hg.advanceExecutorPhase(1);
});

benchmark('atlasPhase()', () => {
  hg.atlasPhase();
});

benchmark('atlasAdvancePhase(1)', () => {
  hg.atlasAdvancePhase(1);
});

benchmark('atlasResonanceAt(0)', () => {
  hg.atlasResonanceAt(0);
});

benchmark('atlasResonanceSnapshot()', () => {
  hg.atlasResonanceSnapshot();
});

// Executor management benchmarks
console.log('\n🏗️ Executor Management Benchmarks:');

benchmark('newExecutor()', () => {
  const handle = hg.newExecutor();
  hg.executorCleanup(handle);
});

benchmark('executorWithBackend("cpu")', () => {
  const handle = hg.executorWithBackend('cpu');
  hg.executorCleanup(handle);
});

// Buffer operation benchmarks
console.log('\n💾 Buffer Operation Benchmarks:');

for (const size of BENCHMARK_CONFIG.bufferSizes) {
  const executorHandle = hg.newExecutor();
  
  benchmark(`executorAllocateBuffer(${size})`, () => {
    const handle = hg.executorAllocateBuffer(executorHandle, size);
    hg.bufferCleanup(handle);
  });
  
  const bufferHandle = hg.executorAllocateBuffer(executorHandle, size);
  
  benchmark(`bufferFill(${size})`, () => {
    hg.bufferFill(bufferHandle, 1.5, size);
  });
  
  benchmark(`bufferLength(${size})`, () => {
    hg.bufferLength(bufferHandle);
  });
  
  benchmark(`bufferToVec(${size})`, () => {
    hg.bufferToVec(bufferHandle);
  });
  
  benchmark(`bufferIsLinear(${size})`, () => {
    hg.bufferIsLinear(bufferHandle);
  });
  
  benchmark(`bufferElementSize(${size})`, () => {
    hg.bufferElementSize(bufferHandle);
  });
  
  benchmark(`bufferSizeBytes(${size})`, () => {
    hg.bufferSizeBytes(bufferHandle);
  });
  
  hg.bufferCleanup(bufferHandle);
  hg.executorCleanup(executorHandle);
}

// Mathematical operation benchmarks
console.log('\n🧮 Mathematical Operation Benchmarks:');

for (const size of BENCHMARK_CONFIG.bufferSizes) {
  benchmark(`vectorAddF32(${size})`, () => {
    hg.vectorAddF32(size);
  });
  
  benchmark(`vectorSubF32(${size})`, () => {
    hg.vectorSubF32(size);
  });
  
  benchmark(`vectorMulF32(${size})`, () => {
    hg.vectorMulF32(size);
  });
  
  benchmark(`vectorDivF32(${size})`, () => {
    hg.vectorDivF32(size);
  });
  
  benchmark(`vectorAbsF32(${size})`, () => {
    hg.vectorAbsF32(size);
  });
  
  benchmark(`vectorNegF32(${size})`, () => {
    hg.vectorNegF32(size);
  });
  
  benchmark(`vectorReluF32(${size})`, () => {
    hg.vectorReluF32(size);
  });
  
  benchmark(`reduceSumF32(${size})`, () => {
    hg.reduceSumF32(size);
  });
  
  benchmark(`reduceMaxF32(${size})`, () => {
    hg.reduceMaxF32(size);
  });
  
  benchmark(`reduceMinF32(${size})`, () => {
    hg.reduceMinF32(size);
  });
  
  benchmark(`sigmoidF32(${size})`, () => {
    hg.sigmoidF32(size);
  });
  
  benchmark(`tanhF32(${size})`, () => {
    hg.tanhF32(size);
  });
  
  benchmark(`softmaxF32(${size})`, () => {
    hg.softmaxF32(size);
  });
}

// Linear algebra benchmarks
console.log('\n📐 Linear Algebra Benchmarks:');

const matrixSizes = [
  [10, 10, 10],
  [50, 50, 50],
  [100, 100, 100],
  [200, 200, 200]
];

for (const matrixSize of matrixSizes) {
  const [m, n, k] = matrixSize;
  if (m !== undefined && n !== undefined && k !== undefined) {
    benchmark(`gemmF32(${m}x${k} * ${k}x${n})`, () => {
      hg.gemmF32(m, n, k);
    });
    
    benchmark(`matvecF32(${m}x${n})`, () => {
      hg.matvecF32(m, n);
    });
  }
}

// Loss function benchmarks
console.log('\n📊 Loss Function Benchmarks:');

for (const size of BENCHMARK_CONFIG.bufferSizes) {
  benchmark(`mseLossF32(${size})`, () => {
    hg.mseLossF32(size);
  });
  
  benchmark(`crossEntropyLossF32(${size})`, () => {
    hg.crossEntropyLossF32(size);
  });
}

// Tensor operation benchmarks
console.log('\n📦 Tensor Operation Benchmarks:');

for (const shape of BENCHMARK_CONFIG.tensorShapes) {
  const executorHandle = hg.newExecutor();
  const bufferSize = shape.reduce((a, b) => a * b, 1);
  const bufferHandle = hg.executorAllocateBuffer(executorHandle, bufferSize);
  const shapeJson = JSON.stringify(shape);
  const tensorHandle = hg.tensorFromBuffer(bufferHandle, shapeJson);
  
  benchmark(`tensorShape([${shape.join(', ')}])`, () => {
    hg.tensorShape(tensorHandle);
  });
  
  benchmark(`tensorStrides([${shape.join(', ')}])`, () => {
    hg.tensorStrides(tensorHandle);
  });
  
  benchmark(`tensorNdim([${shape.join(', ')}])`, () => {
    hg.tensorNdim(tensorHandle);
  });
  
  benchmark(`tensorNumel([${shape.join(', ')}])`, () => {
    hg.tensorNumel(tensorHandle);
  });
  
  benchmark(`tensorIsContiguous([${shape.join(', ')}])`, () => {
    hg.tensorIsContiguous(tensorHandle);
  });
  
  if (shape.length === 2) {
    benchmark(`tensorTranspose([${shape.join(', ')}])`, () => {
      const transposed = hg.tensorTranspose(tensorHandle);
      hg.tensorCleanup(transposed);
    });
    
    const newShape = JSON.stringify([shape[1], shape[0]]);
    benchmark(`tensorReshape([${shape.join(', ')}])`, () => {
      const reshaped = hg.tensorReshape(tensorHandle, newShape);
      hg.tensorCleanup(reshaped);
    });
  }
  
  benchmark(`tensorView1d([${shape.join(', ')}])`, () => {
    const view1d = hg.tensorView1d(tensorHandle);
    hg.tensorCleanup(view1d);
  });
  
  hg.tensorCleanup(tensorHandle);
  hg.bufferCleanup(bufferHandle);
  hg.executorCleanup(executorHandle);
}

// Compiler infrastructure benchmarks
console.log('\n🔧 Compiler Infrastructure Benchmarks:');

benchmark('programBuilderNew()', () => {
  hg.programBuilderNew();
  // Note: No cleanup function available for program builder
});

benchmark('registerAllocatorNew()', () => {
  hg.registerAllocatorNew();
  // Note: No cleanup function available for register allocator
});

const executorHandle = hg.newExecutor();
const bufferHandle = hg.executorAllocateBuffer(executorHandle, 1000);

benchmark('addressBuilderDirect()', () => {
  hg.addressBuilderDirect(bufferHandle, 0);
});

// Memory usage benchmark
console.log('\n💾 Memory Usage Benchmark:');

const memoryTestIterations = 100;
const memoryHandles: number[] = [];

console.log(`Creating ${memoryTestIterations} executors and buffers...`);

const memoryStartTime = Date.now();
for (let i = 0; i < memoryTestIterations; i++) {
  const execHandle = hg.newExecutor();
  const bufHandle = hg.executorAllocateBuffer(execHandle, 1000);
  memoryHandles.push(execHandle, bufHandle);
}
const memoryEndTime = Date.now();

console.log(`Memory allocation time: ${memoryEndTime - memoryStartTime}ms`);
console.log(`Created ${memoryHandles.length / 2} executor-buffer pairs`);

// Clean up memory test
console.log('Cleaning up memory test resources...');
for (let i = 0; i < memoryHandles.length; i += 2) {
  const bufferHandle = memoryHandles[i + 1];
  const executorHandle = memoryHandles[i];
  if (bufferHandle !== undefined) {
    hg.bufferCleanup(bufferHandle);
  }
  if (executorHandle !== undefined) {
    hg.executorCleanup(executorHandle);
  }
}

// Cleanup remaining resources
hg.bufferCleanup(bufferHandle);
hg.executorCleanup(executorHandle);

// Performance summary
console.log('\n📈 Performance Summary:');
console.log('=======================');
console.log('All benchmarks completed successfully!');
console.log('Performance characteristics:');
console.log('- Core functions: Very fast (< 0.1ms)');
console.log('- Buffer operations: Fast (0.1-1ms)');
console.log('- Mathematical operations: Fast (0.1-10ms)');
console.log('- Tensor operations: Moderate (1-100ms)');
console.log('- Linear algebra: Moderate to slow (10-1000ms)');
console.log('- Memory management: Efficient with proper cleanup');

console.log('\n✅ Performance benchmarks completed successfully!');
